OpenRA.Mods.Common/Effects/LineImpactProjectile.cs

```diff
@@ -0,0 +1,271 @@
+
+		[Desc("Beam color in (A,)R,G,B.")]
+		public readonly Color Color = Color.Red;
+
+		[Desc("Beam color is the player's color")]
+		public readonly bool UsePlayerColor = false;
+
+		public IEffect Create(ProjectileArgs args)
+		{
+			var c = UsePlayerColor ? args.SourceActor.Owner.Color.RGB : Color;
+			return new LineImpactProjectile(this, args, c);
+		}
+	}
+
+	public class LineImpactProjectile : IEffect, ISync
+	{
+		readonly LineImpactProjectileInfo info;
+		readonly ProjectileArgs args;
+		[Sync] readonly WDist speed;
+
+		Color color;
+		WAngle angle;
+		WVec offSet;
+
+		[Sync] WPos headPos;
+		[Sync] WPos tailPos;
+		[Sync] WPos target;
+		int length;
+		int facing;
+		[Sync] int headTicks;
+		[Sync] int tailTicks;
+		bool headTravelling, tailTravelling;
+
+		[Sync] public Actor SourceActor { get { return args.SourceActor; } }
+
+		public LineImpactProjectile(LineImpactProjectileInfo info, ProjectileArgs args, Color color)
+		{
+			this.info = info;
+			this.args = args;
+			this.color = color;
+
+			// Both the head and tail start at the source actor, but initially only the head is travelling.
+			headPos = args.Source;
+			tailPos = headPos;
+			headTravelling = true;
+			tailTravelling = false;
+			angle = new WAngle();
+
+			var world = args.SourceActor.World;
+
+			if (info.Speed.Length > 1)
+				speed = new WDist(args.SourceActor.World.SharedRandom.Next(info.Speed[0].Length, info.Speed[1].Length));
+			else
+				speed = info.Speed[0];
+
+			target = args.PassiveTarget;
+			if (info.Inaccuracy.Length > 0)
+			{
+				var inaccuracy = OpenRA.Traits.Util.ApplyPercentageModifiers(info.Inaccuracy.Length, args.InaccuracyModifiers);
+				var maxOffset = inaccuracy * (target - headPos).Length / args.Weapon.Range.Length;
+				target += WVec.FromPDF(world.SharedRandom, 2) * maxOffset / 1024;
+			}
+
+			facing = OpenRA.Traits.Util.GetFacing(target - headPos, 0);
+
+			// Update the target position with the range we shoot beyond the target by
+			// I.e. we can deliberately overshoot, so aim for that position
+			var dir = new WVec(0, -1024, 0).Rotate(WRot.FromFacing(facing));
+			var overshoot = dir * info.BeyondTargetRange.Length / 1024;
+			target = target + overshoot;
+
+			length = Math.Max((target - headPos).Length / speed.Length, 1);
+
+			if (args.SourceActor.IsInWorld)
+				offSet = args.Source - args.SourceActor.CenterPosition;
+		}
+
+		public void Tick(World world)
+		{
+			headTicks++;
+			if (headTicks >= length)
+			{
+				headPos = target;
+				headTravelling = false;
+			}
+			else
+			{
+				if (headTravelling)
+					headPos = WPos.LerpQuadratic(args.Source, target, angle, headTicks, length);
+			}
+
+			// While the head is travelling, the tail must start to follow BeamLengthDuration ticks later.
+			// Alternatively, also stop emitting the beam if source actor dies or is ordered to stop.
+			if ((headTicks >= info.BeamFireDuration && !tailTravelling) || args.SourceActor.IsDead || args.SourceActor.IsIdle)
```

Comment:
Putting in an `!args.SourceActor.Trait<AttackBase>().IsAttacking` instead of `args.SourceActor.IsIdle` should work better when ordering the turreted firerers to move. And probably something should be done in the case one attack activity gets replaced by another (either stop the current beam or change target position once moving targets are supported).